wave_optimization/include/wave/optimization/factor_graph/factor_graph.hpp

```diff
@@ -0,0 +1,82 @@
+#ifndef WAVE_OPTIMIZATION_FACTOR_GRAPH_FACTOR_GRAPH_HPP
+#define WAVE_OPTIMIZATION_FACTOR_GRAPH_FACTOR_GRAPH_HPP
+
+#include <iostream>
+#include <fstream>
+#include <vector>
+#include <map>
+#include <set>
+#include <memory>
+
+#include "wave/utils/utils.hpp"
+#include "wave/optimization/factor_graph/factor.hpp"
+#include "wave/optimization/factor_graph/variable.hpp"
+
+namespace wave {
+
+class FactorGraph {
+ public:
+    std::multimap<std::shared_ptr<Variable>, std::shared_ptr<Factor>> graph;
+    std::set<std::shared_ptr<Variable>> variables;
+    std::vector<std::shared_ptr<Factor>> factors;
+
+    FactorGraph() : graph{}, variables{}, factors{} {}
```

Comment:
It just gets out of hand if you have many members or multiple constructors.

If paranoid you can use a default member initializer:
```cpp
std::vector<std::shared_ptr<Factor>> factors{};
```